firstpartydata/first_party_data.go

```diff
@@ -0,0 +1,376 @@
+package firstpartydata
+
+import (
+	"fmt"
+	"github.com/evanphx/json-patch"
+	"github.com/mxmCherry/openrtb/v15/openrtb2"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"github.com/prebid/prebid-server/util/jsonutil"
+)
+
+const (
+	siteKey = "site"
+	appKey  = "app"
+	userKey = "user"
+	dataKey = "data"
+
+	userDataKey        = "userData"
+	appContentDataKey  = "appContentData"
+	siteContentDataKey = "siteContentData"
+)
+
+func ExtractGlobalFPDData(req *openrtb_ext.RequestWrapper) (map[string][]byte, error) {
+	//If {site,app,user}.ext.data exists, collect it and remove {site,app,user}.ext.data
+
+	fpdReqData := make(map[string][]byte, 3)
+
+	siteExt, err := req.GetSiteExt()
+	if err != nil {
+		return nil, err
+	}
+
+	if siteExt != nil && len(siteExt.GetExt()[dataKey]) > 0 {
+		fpdReqData[siteKey] = siteExt.GetExt()[dataKey]
+		newSiteExt, err := jsonutil.DropElement(req.Site.Ext, dataKey)
+		if err != nil {
+			return nil, err
+		}
+		req.Site.Ext = newSiteExt
```

Comment:
@hhhjort Should this (along with app and user) be using `SetExt` to properly set the modified flag?

Reply:
Refactored using request wrapper